alf/networks/disjoint_policy_value_network.py

```diff
@@ -0,0 +1,263 @@
+            input_tensor_spec=observation_spec,
+            kernel_initializer=kernel_initializer)
+
+        # +------------------------------------------+
+        # | Step 2: Projection for the policy branch |
+        # +------------------------------------------+
+
+        projection_net = _create_projection_net_based_on_action_spec(
+            discrete_projection_net_ctor=discrete_projection_net_ctor,
+            continuous_projection_net_ctor=continuous_projection_net_ctor,
+            input_size=self._actor_encoder.output_spec.shape[0],
+            action_spec=action_spec)
+
+        if alf.nest.is_nested(projection_net):
+            # In the case when the projection net is multi-fold (because actor
+            # spec is a nest), force picking up the parameters inside those
+            # project networks into this DisjointPolicyValueNetwork instance.
+            self._projection_net_module_list = torch.nn.ModuleList(
+                alf.nest.flatten(projection_net))
+            self._actor_branch = alf.nn.Sequential(
+                self._actor_encoder,
+                alf.nn.Branch(projection_net, name='NestedProjection'))
+        else:
+            self._actor_branch = alf.nn.Sequential(self._actor_encoder,
+                                                   projection_net)
+
+        # +------------------------------------------+
+        # | Step 3: Value head of the aux branch     |
+        # +------------------------------------------+
+
+        # Note that the aux branch value head belongs to the policy component.
+
+        # Like the value head Aux head is outputing value estimation
+        self._aux_branch = alf.nn.Sequential(
+            self._actor_encoder,
+            alf.layers.FC(
+                input_size=self._actor_encoder.output_spec.shape[0],
+                output_size=1), alf.layers.Reshape(shape=()))
+
+        # +------------------------------------------+
+        # | Step 4: Value head of the value branch   |
+        # +------------------------------------------+
+
+        if is_sharing_encoder:
+            # Use the same encoder, but the encoder is DETACHED.
+            self._value_branch = alf.nn.Sequential(
+                self._actor_encoder, alf.layers.Detach(),
+                alf.layers.FC(
+                    input_size=self._actor_encoder.output_spec.shape[0],
+                    output_size=1), alf.layers.Reshape(shape=()))
+        else:
+            # If not sharing encoder, create a separate encoder for the value
+            # component.
+            value_encoder = encoding_network_ctor(
+                input_tensor_spec=observation_spec,
+                kernel_initializer=kernel_initializer)
+
+            self._value_branch = alf.nn.Sequential(
+                value_encoder,
+                alf.layers.FC(
+                    input_size=self._actor_encoder.output_spec.shape[0],
+                    output_size=1), alf.layers.Reshape(shape=()))
```

Comment:
Have you considered putting every thing in a single container? That might simplify the code.

Reply:
Yep I was thinking about using one `nn.Branch`, but hesitated because in certain cases we might want to run just one head. But given that the head is usually very lite, maybe putting them under one `nn.Branch` is better. Also good for readability.

Reply:
Now combined into one container.